clients/src/test/java/org/apache/kafka/common/utils/UtilsTest.java

```diff
@@ -158,6 +161,39 @@ public void testCloseAll() {
         }
     }
 
+    @Test
+    public void testReadFully() throws IOException {
+        RandomAccessFile raf = new RandomAccessFile(TestUtils.tempFile(), "rw");
```

Comment:
Correct me if I am wrong. Remember readFully has two predicates that are used to check if we can exit the loop: 1. FileChannel.read returns -1; 2. buffer.hasRemaining = false.

I already wrote a unit test case named 'testReadFullyThrowIOExceptionForEOF' to verify readFully can throw IOException when encountering EOF. That covered the first predicate.

Let's see the second predicate: buffer.hasRemaining(). If covering this check, seems I have to use a mocked byte buffer. In doing so can I mock the hasRemaining's return value. Otherwise, a real byte buffer always has remaining since a mocked file channel does no operation to the buffer. 

Do I make myself clear?